### PR TITLE
[SPARK-37496][SQL] Migrate ReplaceTableAsSelectStatement to v2 command

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
This PR migrates `ReplaceTableAsSelectStatement` to the v2 command

### Why are the changes needed?
Migrate to the standard V2 framework

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
existing tests

### CODE REVIEW

**Comment by huaxingao** [comment]
cc @cloud-fan Could you please take a look?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 1]
```diff
@@ -586,19 +586,13 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
-          None,
-          Map.empty,
-          Some(source),
-          Map.empty,
-          extraOptions.get("path"),
-          extraOptions.get(TableCatalog.PROP_COMMENT),
-          extraOptions.toMap,
-          None,
-          orCreate = true)      // Create the table if it doesn't exist
+        val tableSpec = TableSpec(None, Map.empty, Some(source), Map.empty,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:589]: let's use named parameter here: `TableSpec(bucketSpec = None, ...)`

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriterV2.scala [hunk 2]
```diff
@@ -195,20 +195,12 @@ final class DataFrameWriterV2[T] private[sql](table: String, ds: Dataset[T])
...
-        None,
-        properties.toMap,
-        provider,
-        Map.empty,
-        None,
-        None,
-        options.toMap,
-        None,
-        orCreate = orCreate))
+    val tableSpec = TableSpec(None, properties.toMap, provider, Map.empty,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriterV2.scala:198]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala [hunk 3]
```diff
@@ -196,11 +204,19 @@ case class AtomicReplaceTableAsSelectExec(
...
     plan: LogicalPlan,
     query: SparkPlan,
-    properties: Map[String, String],
+    tableSpec: TableSpec,
     writeOptions: CaseInsensitiveStringMap,
     orCreate: Boolean,
     invalidateCache: (TableCatalog, Table, Identifier) => Unit) extends TableWriteExecHelper {
 
+  val properties = {
+    val props = CatalogV2Util.convertTableProperties(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala:213]: nit: should the util function take `TableSpec` directly? and also fill the ownership property. Then we can simplify the code here and other places.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 1 above)
Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:589]: Fixed

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala (hunk 3 above)
Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala]: Fixed. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 4]
```diff
@@ -586,19 +586,22 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
+          bucketSpec = None,
+          properties = Map.empty,
+          provider = Some(source),
+          options = Map.empty,
+          location = extraOptions.get("path"),
+          comment = extraOptions.get(TableCatalog.PROP_COMMENT),
+          serde = None,
+          external = false)
+        ReplaceTableAsSelect(
+          UnresolvedDBObjectName(nameParts, isNamespace = false),
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:599]: shall we prepend `catalog.name`?

Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:599]: Seems no need? I stepped into one of the test, the `nameParts` here is "testcat", "ns", "t", so it already contains `catalog.name`.

**Review by cloud-fan — APPROVED** [review]

**Comment by huaxingao** [comment]
Merged to master. Thank you very much for reviewing! @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 5]
```diff
@@ -586,19 +586,22 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
-          Some(source),
-          Map.empty,
-          extraOptions.get("path"),
-          extraOptions.get(TableCatalog.PROP_COMMENT),
-          extraOptions.toMap,
-          None,
-          orCreate = true)      // Create the table if it doesn't exist
+          df.queryExecution.analyzed,
+          tableSpec,
+          writeOptions = Map.empty,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:603]: I think there is a bug here. Previously we pass `extraOptions.toMap` as the write options, now we don't. @huaxingao can you help to fix it with a test case? thanks!

Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:603]: Sorry for the bug. Will fix.

**Pull request closed without merging**